#10402 fix theme templates overrides in environment emulation cross-area

alessandroniciforo opened the pull request:
Row 141 and 145 are obviously returning a theme id, so we'd better load the theme by id. I'm not sure if execution will fall into row 143 what to expect.

### Fixed Issues (if relevant)
<!--- Provide a list of fixed issues in the format magento/magento2#<issue_number>, if relevant  -->
1. magento/magento2#10402: Theme templates overrides don't work in environment emulation cross-area

alessandroniciforo committed:
#10402 fix theme templates overrides in environment emulation cross-area

Row 141 and 145 are obviously returning a theme id, so we'd better load the theme by id. I'm not sure if execution will fall into row 143 what to expect.

orlangur commented:
> I'm not sure if execution will fall into row 143 what to expect.

In such case `$theme` will not contain `themeId` and suggested code will fail :) While original one worked...

orlangur commented:
Note that `magento:2.1-develop` target branch should be used for backports of pull requests already merged to `develop` branch only.

orlangur commented on lib/internal/Magento/Framework/View/Asset/Repository.php:
```diff
@@ -146,7 +146,7 @@ public function updateDesignParams(array &$params)
 
         if ($theme) {
-            $params['themeModel'] = $this->getThemeProvider()->getThemeByFullPath($area . '/' . $theme);
+            $params['themeModel'] = $this->getThemeProvider()->getThemeById($theme);
```
Please rewrite method in such a way that if we go through line 141 `getThemeById` call is used and `getThemeByFullPath` otherwise.

Also, such case with emulated cross-area needs to be covered with integration test.

alessandroniciforo replied:
Actually, also for line 145 `getThemeById` should be called as `getConfigurationDesignTheme` returns the id. I don't have time to write integration test at this moment, unfortunately.
Should I open another PR on develop branch?

orlangur replied:
I changed base branch, please force push into the same branch applying your commit on top of `develop`.

Yeah, you are right about line 145.

Just finish your implementation without integration test then.

alessandroniciforo committed:
#10402 fix theme templates overrides in environment emulation cross-area

 Load theme by full path only in case we received a 'theme' in $params

alessandroniciforo committed:
#10402 fix unit tests to cover different cases for updateDesignParams

alessandroniciforo commented:
Pushed the suggested change and the fix to the unit test.

alessandroniciforo committed:
fix unit test code style

orlangur commented:
Please try to simplify modified method
```
1) Magento\Test\Php\LiveCodeTest::testCodeMess
PHP Code Mess has found error(s):
/home/travis/build/magento/magento2/lib/internal/Magento/Framework/View/Asset/Repository.php:130	The method updateDesignParams() has an NPath complexity of 240. The configured NPath complexity threshold is 200.
Failed asserting that 2 matches expected 0.
```

Maybe we can leave only one exception throwing and decrease amount of `if`s?

alessandroniciforo committed:
#10402 ensure if the theme is loaded by id that it belongs to the requested area

alessandroniciforo committed:
#10402 reduce npath complexity

alessandroniciforo committed:
#10402 refactoring to reduce method complexity and to increase readability

alessandroniciforo commented:
Just moving the exception throwing was not enough to go below 200. On top of that, I found that in case we load the theme by id we should ensure that the theme exists in the requested area (this is implicit in the load by full path), so I added that check and refactored a bit.
Hopefully it looks better now.

alessandroniciforo committed:
#10402 change implementation to detect automatically whether the theme param returned by getConfigurationDesignTheme is an id or a path

alessandroniciforo commented:
The issue unveiled by the integration tests run is that `Magento\Theme\Model\View::getConfigurationDesignTheme()` can either return:

- theme id: if execution pass into rows [#174](https://github.com/magento/magento2/blob/develop/app/code/Magento/Theme/Model/View/Design.php#L174) or [#179](https://github.com/magento/magento2/blob/develop/app/code/Magento/Theme/Model/View/Design.php#L179) AND if the design configuration has been saved on the backend, thus inserting the config line into `core_config_data`
- theme path: in all the other cases. There's a fallback at row [#188](https://github.com/magento/magento2/blob/develop/app/code/Magento/Theme/Model/View/Design.php#L188) so that at worst it takes the theme passed in its constructor, injected in its [di.xml](https://github.com/magento/magento2/blob/develop/app/code/Magento/Theme/etc/di.xml#L62), which is a theme path

So my (ugly) solution is that, in case we rely on `Magento\Theme\Model\View::getConfigurationDesignTheme()`, we should detect whether the passed parameter is an id or a path.

A more appropriate solution would be to modify the implementation of `getConfigurationDesignTheme` to have a mean of knowing whether the returned value is an id or a path, but I'm not confident enough to refactoring that part too.

orlangur commented on lib/internal/Magento/Framework/View/Asset/Repository.php:
```diff
@@ -135,22 +135,17 @@ public function updateDesignParams(array &$params)
-            }
-        } elseif (empty($params['themeModel'])) {
+        if (empty($params['themeModel'])) {
             $params['themeModel'] = $this->getDefaultParameter('themeModel');
```
For better readability please do `$themeModel = ...` instead of `$params['themeModel'] = ...` and only in the very end do `$params['themeModel'] = $themeModel ?: $this->getDefaultParameter('themeModel')`.

orlangur commented on lib/internal/Magento/Framework/View/Asset/Repository.php:
```diff
@@ -167,6 +162,52 @@ public function updateDesignParams(array &$params)
+     */
+    private function getTheme($theme, $area)
+    {
+        if (is_numeric($theme)) {
```
Yeah, looking into `\Magento\Theme\Model\View\Design::getConfigurationDesignTheme` this seems to be the best option to me (although we won't support string theme identifiers consisting of digits only).

Can we eliminate this method and still pass PHPMD? `getThemeById` and `getThemeByPath` seem much more understandable to me at the place of call but I would like to avoid method named like `getThemeByIdOrPath`.

alessandroniciforo replied:
Moving this in the method we wouldn't pass the cyclomatic complexity

orlangur commented on lib/internal/Magento/Framework/View/Asset/Repository.php:
```diff
@@ -167,6 +162,52 @@ public function updateDesignParams(array &$params)
+    {
+        $themeModel = $this->getThemeProvider()->getThemeByFullPath($area . '/' . $themePath);
+        if (!$themeModel) {
+            throw new \UnexpectedValueException("Could not find theme '$themePath' for area '$area'");
```
Please reword `Could not find theme by full path ...`

orlangur commented on lib/internal/Magento/Framework/View/Asset/Repository.php:
```diff
@@ -167,6 +162,52 @@ public function updateDesignParams(array &$params)
+    {
+        $themeModel = $this->getThemeProvider()->getThemeById($themeId);
+        if (!$themeModel || $themeModel->getArea() != $area) {
+            throw new \UnexpectedValueException("Could not find theme '$themeId' for area '$area'");
```
Please reword `Could not find theme by id ...`

orlangur commented on lib/internal/Magento/Framework/View/Test/Unit/Asset/RepositoryTest.php:
```diff
@@ -157,36 +157,218 @@ public function testUpdateDesignParamsWrongTheme()
+        $this->themeProvider
+            ->expects($this->any())
+            ->method('getThemeByFullPath')
+            ->with('frontend/Magento/luma')
```
Please avoid Luma theme mentioning in framework, just use some fictional theme path (fix both occurrences).

orlangur commented on lib/internal/Magento/Framework/View/Asset/Repository.php:
```diff
@@ -135,22 +135,17 @@ public function updateDesignParams(array &$params)
         }
 
```
As this method was refactored significantly, please remove `@SuppressWarnings(PHPMD.CyclomaticComplexity)`, probably it is already passing this check.

orlangur commented on lib/internal/Magento/Framework/View/Asset/Repository.php:
```diff
@@ -167,6 +162,52 @@ public function updateDesignParams(array &$params)
+    }
+
+    /**
+     * @param string $themeId
```
Change to `int`.

orlangur commented:
Thanks, your findings look great to me, especially that there was `theme_id` in some tests not containing number and that there are actually some integration and functional tests enforcing this logic.

Would be great if you add case to check new logic into some existing integration test but it's not mandatory. Please check my comments to slightly polish up the code.

alessandroniciforo committed:
#10402 cleanup

orlangur commented:
@alessandroniciforo there is one real failure left: https://travis-ci.org/magento/magento2/jobs/261023165
```
There was 1 failure:
1) Magento\Deploy\DeployTest::testDeploy
Failed asserting that 3266.36328125 is less than 1177.6.
/home/travis/build/magento/magento2/dev/tests/integration/testsuite/Magento/Deploy/DeployTest.php:232
/home/travis/build/magento/magento2/dev/tests/integration/testsuite/Magento/Deploy/DeployTest.php:151
                                                                        
FAILURES!                                                               
Tests: 1291, Assertions: 5716, Failures: 1, Incomplete: 43, Skipped: 11.
```

Could you please look into it?

ishakhsuvarov commented:
@alessandroniciforo @orlangur Is there any progress on this Pull Request?

orlangur commented:
@ishakhsuvarov I planned to look into this failure if Alessandro will not do it as fix seems quite valuable to me. Didn't have time yet, will have during this weekend.

Feel free to investigate failure yourself (maybe it occurred already in some other circumstances already?) or close this PR for now, I'll reopen when ready.

okorshenko commented:
Hi All 
We are closing this PR due to inactivity. If you would like to proceed with this PR, please reopen it any time

diamondavocado commented:
@orlangur Could you please re-open this PR? It was very close to being finished and it would solve issues https://github.com/magento/magento2/issues/10402 and https://github.com/magento/magento2/issues/8884.

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=10403) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=10403) before we can accept your contribution.<br/>

alessandroniciforo commented:
Hey guys,
sorry but I couldn't understand the reason for the failure of the integration test and I've since moved away from Magento.
Looking at the PR now, it seems that the issue has been solved already on 2.3-develop by this commit https://github.com/magento/magento2/commit/d3e32b4e5db80f7c04734f63cfeec83155613222.
I don't know if it makes sense to merge it to 2.1 (which I don't know if it's still maintained) or to 2.2

sidolov commented:
Hi @alessandroniciforo , thank you for your contribution! Unfortunately, we are no longer accept contributions to 2.1-develop and 2.2-develop branches. Looks like the issue was fixed in the scope of d3e32b4e5db80f7c04734f63cfeec83155613222, as you mentioned before. 
I'm closing this PR as not relevant. 
Thank you!
